[REEF-1557] Refactor State, EvaluatorState, and related classes.

motus opened the pull request:
All changes are for readability and maintainability of the code. There are no changes in logic.

Summary:

  * Move all state-checking logic into corresponding enums.
  * Make state atomic in `EvaluatorStatusManager`.
  * Refactor `EvaluatorManager` for readability.
  * Add javadocs and cleanup the code.

JIRA: [REEF-1557](https://issues.apache.org/jira/browse/REEF-1557)

tcNickolas commented on lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/EvaluatorRestartState.java:
```diff
@@ -65,32 +66,50 @@
    */
-  public static boolean isLegalTransition(final EvaluatorRestartState from, final EvaluatorRestartState to) {
-    switch(from) {
+  public static boolean isLegalTransition(
```
Needs `@Deprecated` annotation to match `@deprecated` in comment

tcNickolas commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/DriverStatus.java:
```diff
@@ -22,9 +22,57 @@
+      }
+
+    default:
+      return false;
```
This is a logic change, even though minor - previously this switch checked all states we consider and threw exception if current state is unconsidered. This is potentially useful - if we add another state we'll be forced to revisit this logic. I'd keep it

tcNickolas commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/DriverStatusManager.java:
```diff
@@ -152,10 +114,12 @@ public synchronized void onError(final Throwable exception) {
+    if (this.isClosing()) {
       LOG.log(Level.WARNING, "Received an exception while already in shutdown.", exception);
     } else {
+
```
I can't find which coding style requires empty line before first statement of the block.

tcNickolas commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/DriverStatusManager.java:
```diff
@@ -234,21 +201,33 @@ public synchronized void sendJobEndingMessageToClient(final Optional<Throwable>
+   * Check if the driver is in process of shutting down (either gracefully or due to an error).
+   * @return true if the driver is shutting down (gracefully or otherwise).
+   * @deprecated TODO[JIRA REEF-1560] Use isClosing() method instead. Remove after version 0.16
+   */
```
Needs `@Deprecated` annotation

motus commented:
Done with the fixes - thanks for reviewing @tcNickolas !

tcNickolas commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorState.java:
```diff
@@ -22,17 +22,150 @@
+
+  /**
+   * Check if evaluator is in the initial state (ALLOCATED).
+   * @return true if allocated, false otherwise.
```
Should `allocated` here be uppercase, as name of state? Same question for `submitted` in next function

tcNickolas commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorState.java:
```diff
@@ -22,17 +22,150 @@
+      }
+
+    default:
+      return false;
```
Behavior change, doesn't check for all considered states and throw by default now

tcNickolas commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorStatusManager.java:
```diff
@@ -31,160 +32,159 @@
+   * @throws IllegalStateException if state transition is not valid.
+   */
+  private void setState(final EvaluatorState toState) {
+    for (;;) {
```
Intellij recommends to use while loop instead

tcNickolas commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/pojos/State.java:
```diff
@@ -21,20 +21,128 @@
+    return this == RUNNING;
+  }
+
+
```
Extra empty line

tcNickolas commented:
I've done a pass. Two overall comments:

1. We need a unified policy regarding empty lines. Even within this change sometimes you add more empty lines, and sometimes you remove existing ones (in `EvaluatorStatusPOJO` constructor). For example, I'm not sure we need empty lines after opening bracket

2. Same regarding `this.` prefix for instance variables and methods. I'm not convinced we need to start adding `this.` in cases in which instance variable/method is not shadowed by other names. Corresponding Checkstyle [check](http://checkstyle.sourceforge.net/config_coding.html#RequireThis) warns that it is controversial. I've tried running it on our code with `validateOnlyOverlapping=true`, and it found no violations, and Intellij highlights all added `this.` as not needed.

motus commented:
@tcNickolas thanks! I've fixed the code and squashed the last commits. Regarding the code style:

  1. I think we need blank lines to improve readability. e.g. I sometimes add blank lines to separate final member variables from the mutable ones, or use them between logical blocks of code - say, to separate logging from main logic, and so on. I think we can be a bit creative here as long as checkstyle does not complain.
  2. I agree that `this.` prefix is often redundant. We use that style from the beginning of the project because it was in some coding guidelines that we had followed then (or was it some default IDE setting?) Anyway, we are stuck with it now, so I use it to be consistent with the rest of the code. I absolutely don't mind if someone would offer a PR that converts all code to `this`-less style.

jwang98052 commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java:
```diff
@@ -111,17 +108,19 @@
 
   @Inject
   private EvaluatorManager(
+      @Parameter(EvaluatorIdentifier.class) final String evaluatorId,
```
Is the convention to put the named parameters at beginning?

motus replied:
@jwang98052 I don't think we have it written anywhere, but we should - IMHO having `@Parameter` annotations first and keeping them together adds to the code readability.

jwang98052 commented on lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorManager.java:
```diff
@@ -182,28 +182,27 @@ public static String getJobIdentifier() {
+
     } else {
-      LOG.log(Level.WARNING, "Evaluator allocated event fired more than once.");
+      LOG.log(Level.WARNING, "EvaluatorAllocated event fired more than once.");
```
The event name is AllocatedEvaluator

markusweimer commented:
Regarding `this.`: This is on me :) I've just always written code like that. But with current IDEs, it really is redundant as they highlight object attributes differently from local variables. Still, it makes the code easier to parse & review without a proper IDE, IMHO, e.g. on the code review page on GitHub.

tcNickolas commented:
I think when I was looking at Checkstyle we had plenty of places in code in which we don't prefix members with `.this`. I suggest we take `.this` question to the mailing list and vote on a single style, and then configure a check to enforce it.
As for doing code review on Github, I often checkout pull request and view it in proper IDE to simplify the process :-)

motus commented:
So how about we merge this PR and open a low-priority JIRA to state whatever style preferences we want eventually to converge to? Besides `this.`, we can also mention the ordering of `@Parameter`-annotated arguments, and other such things.

tcNickolas commented:
I'm fine with that. @jwang98052 do you have any more comments?

jwang98052 commented:
LGTM

tcNickolas commented:
I'll test and merge
